File: Heap/maxHeap.js
/*

  1.  CREATE A FUNCTION THAT WILL CREATE NODES FOR HEAP CLASS NAMED Node
  2.  CREATE A MAX BINARY HEAP CLASS NAMED MaxBH
  3.  CREATE A METHOD TO INSERT NODES INTO THE MAX HEAP CLASS NAMED insert
  4.  CHREATE A HELPER FUNCTION TO SORT HEAP INTO PROPER ORDER AFTER INSERTING A NEW NODE NAMED insert
        MAX HEAP: PARENT NODE'S VALUE IS ALWAYS GREATER THAN CHILD NODE'S VALUE
        MIN HEAP: PARENT NODE'S VALUE IS ALWAYS LESS THAN THE CHILD NODE'S VALUE
  5.  CREATE A METHO TO REMOVE THE MAX NODE FROM THE MAX BINARY HEAP TREE
  6.  CREATE A HELPER FUNCTION TO SORT HEAP INTO PROPER ORDER AFTER MAX NODE IS REMOVED NAMED remove

*/

const Node = function (val) {
  this.val = val;
};

const MaxBH = class {
  constructor() {
    this.root = [];
  }

  insert(val) {
    const incoming = new Node(val);
    if (this.root) {
      this.root[this.root.length] = incoming;
      let cIdx = this.root.length - 1;
      let pIdx = Math.floor((cIdx - 1) / 2);
      this.heapUp(cIdx, pIdx);
    } else this.root = incoming;
    return this;
  }

  heapUp(cIdx, pIdx) {
    let tmp;
    while (this.root[pIdx] && this.root[pIdx].val < this.root[cIdx].val) {
      tmp = this.root[pIdx];
      this.root[pIdx] = this.root[cIdx];
      this.root[cIdx] = tmp;
      cIdx = pIdx;
      pIdx = Math.floor((cIdx - 1) / 2);
    }
  }

  remove() {
    if (this.root) {
      const max = this.root[0];
      if (this.root.length > 1) {
        this.root[0] = this.root.pop();
        this.heapDown();
      } else {
        this.root.pop();
      }
      return max;
    }
  }

  heapDown() {
    let pIdx = 0,
      leftCIdx = 1,
      rightCIdx = 2,
      bigCIdx,
      tmp;

    while (
      (this.root[leftCIdx] && this.root[pIdx].val < this.root[leftCIdx].val) ||
      (this.root[rightCIdx] && this.root[pIdx].val < this.root[rightCIdx].val)
    ) {
      if (!this.root[rightCIdx]) bigCIdx = leftCIdx;
      else
        bigCIdx =
          this.root[leftCIdx].val > this.root[rightCIdx].val
            ? leftCIdx
            : rightCIdx;

      tmp = this.root[bigCIdx];
      this.root[bigCIdx] = this.root[pIdx];
      this.root[pIdx] = tmp;
      pIdx = bigCIdx;
      leftCIdx = pIdx * 2 + 1;
      rightCIdx = leftCIdx + 1;
    }
  }

  printer() {
    let head = this.root;
    let one = 'one';
  }
};

const myMaxBH = new MaxBH();
// myMaxBH.insert(41);
// myMaxBH.insert(39);
// myMaxBH.insert(33);
// myMaxBH.insert(18);
// myMaxBH.insert(27);
// myMaxBH.insert(12);
// myMaxBH.insert(55);

myMaxBH.insert(5);
myMaxBH.insert(10);
myMaxBH.insert(15);
myMaxBH.insert(20);
myMaxBH.insert(25);
myMaxBH.insert(30);
myMaxBH.insert(8);

console.log(myMaxBH.root, '\n\n');

console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());
console.log("remove method's return value: ", myMaxBH.remove());

myMaxBH.printer();
// console.log("remove method's return value: ", myMaxBH.remove(), '\n\n');
// myMaxBH.insert(100);

// console.log(myMaxBH.root, '\n\n');

File: Heap/max_heap.py
"""
  1.  Create Node class that will provide nodes for the MBH class
      I.  It will take a value and make it a property
"""
##################################   NODE CLASS   ##################################################
class Node:
  def __init__(self, val, data):
    self.val = val
    self.data = data


"""
  1.  CREATE A MAX BINARY HEAP CLASS THAT WILL SERVE AS A MAX BINARY HEAP DATA STRUCTURE.
      METHODS:
        1.  insert_node will insert a new node into the max binary heap and move it into
            its proper place.
"""


class Max_Binary_Heap:
  def __init__(self):
    self.values = []

  ##################################################################################################
  ####################       METHODS TO INSERT A NODE INTO MBH       ###############################
  ##################################################################################################


   ##################            _heap_up METHOD               #####################################
  #   time complexity: O(n)

  """
    1.  CREATE A CHILD INDEX FROM THE LAST INDEX IN MBH AND USE IT TO CREATE PARENT INDEX
    2.  loop while parent index is zero or greater and the value at parent
        index is smaller than the value at child's index
          I.      swap the nodes at those two indices
          II.     adjust the two indices for the next iteration
  """

  def _heap_up(self):
    c_Idx = len(self.values) - 1
    p_Idx = (c_Idx - 1)//2
    while p_Idx >= 0 and self.values[p_Idx].val < self.values[c_Idx].val:
      self.values[p_Idx], self.values[c_Idx] = self.values[c_Idx], self.values[p_Idx]
      c_Idx = p_Idx
      p_Idx = (c_Idx - 1)//2


  ##################   insert_node METHOD     ######################################################
  #   time complexity: O(n)

  """
    1.  IF BOTH ARGUMENTS ARE VALID
        I.    it creates a new node, call it incoming_node
        II.   IF THERE IS ATLEAST A SINGLE NODE IN MBH
              a.    add the incoming node to the end of the MBH
              b.    call helper method named _heap_up
        III.  ELSE ADD THE NODE TO THE MBH
        IV.   RETURN SELF
  """


  def insert_node(self, val, data):
    if type(val) == int and type(data) == str:
      incoming = Node(val, data)
      if len(self.values):
        self.values.append(incoming)
        self._heap_up()
      else:
        self.values.append(incoming)
    return self




  ##################################################################################################
  ####################       METHODS TO REMOVE A NODE INTO MBH       ###############################
  ##################################################################################################




  ###################        _heap_down METHOD         #############################################
  #   time complexity: O(n)

  """
    1.    create three indices: parent index, left child index, and a right child index
    2.    loop while the node's value at the parent index is smaller than
          either child's value at their respecitive indices in the MBH and while child
          indices are not longer than the length of the MBH
          I.    find the child node with the larger value
          II.   swap the two nodes at those indices in the MBH
          III.  make the parent index the larger child's index and calculate the
                children's indices
  """

  def _heap_down(self):
    # create necessary indices for swapping  #####################################################
    p_Idx = 0
    l_child_Idx = 1
    r_child_Idx = 2
    #  loop while either child's index is within reasonable range and that child's value is larger than the parent's value
    while (l_child_Idx < len(self.values) and self.values[p_Idx].val < self.values[l_child_Idx].val) or (r_child_Idx < len(self.values) and self.values[p_Idx].val < self.values[r_child_Idx].val):
      #  find the larger child  ##################################################################
      larger_child_Idx = l_child_Idx
      if r_child_Idx < len(self.values) and self.values[l_child_Idx].val < self.values[r_child_Idx].val:
        larger_child_Idx = r_child_Idx
      #  swap parent with larger child  ############################################################
      self.values[p_Idx], self.values[larger_child_Idx] = self.values[larger_child_Idx], self.values[p_Idx]
      #  adjust indices for next iteration  ######################################################
      p_Idx = larger_child_Idx
      l_child_Idx = p_Idx * 2 + 1
      r_child_Idx = l_child_Idx + 1



  ########################        shift METHOD         #############################################
  #   time complexity: O(n)

    """
      1.  IF THERE IS AT LEAST ONE NODE IN THE MBH
          I.    create a variable, call it max_node, and assign it the value of the first node in MBH
          II.   if lenght of MBH is one
                a.  pop off the only node in MBH
          III.  else MBH has more than one node
                a.  assign the first node of the MBH the value of the last node in MBH
                b.  call the _heap_down helper funciton
          IV.   return max_node
    """

  def shift(self):
    #  IF THERE IS AT LEAST ONE NODE IN MBH    #####################################################
    if len(self.values):
      max_node = self.values[0]
      if len(self.values) == 1:
        self.values.pop()
      else:
        self.values[0] = self.values.pop()
        self._heap_down()
      return max_node



  ##################################   for testing    ##############################################
  def printer(self):
    print(self.values)



my_MBH = Max_Binary_Heap()
my_MBH.insert_node(5, 'oranges')
my_MBH.insert_node(10, 'bananas')
my_MBH.insert_node(15, 'apples')
my_MBH.insert_node(20, 'avocados')
my_MBH.insert_node(25, 'tomatos')
my_MBH.insert_node(30, 'lettuce')
my_MBH.insert_node(8, 'onions')

print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift().val)
print('remove max node from MBH: ', my_MBH.shift())
my_MBH.printer()

